### ISSUE TITLE
"Malformed PASV response" when DataConnectionType was not set

### ISSUE THREAD

**Issue by tekhedd** [hgupta9/FluentFTP#95]
I am seeing this error using FTP(S)
 * vsftpd server,
 * NuGet build 16.4.0
 * .NET runtime on Windows 7

(Can not reproduce under mono)

  FluentFTP.FtpException: Malformed PASV response: Entering Extended Passive Mode (|||50027|).
     at FluentFTP.FtpClient.OpenPassiveDataStream(FtpDataConnectionType type, String command, Int64 restart)

This is an EPSV response, so the error message is trivially correct: it is definitely not a valid PASV response. However, I have not set "DataConnectionType", so presumably it should be set to the default AutoPassive setting. Judging from the FtpClient.cs code, it should not even be possible to reach this error message with DataConnectionType == AutoPassive, so perhaps this is not the default as documented? Or some bit of the code is changing "type" after requesting EPSV mode?

Or, possibly there is a bug in vsftpd that is attempting to set EPSV when PASV is requested, although that seems unlikely. 

I will dig deeper. My debug-trace listener does not seem to be engaging on the production build, so this might take a while.

**Comment by tekhedd** [comment]
Code review suggests that this may be a result of unexpected/unread messages in the socket, resulting in an immediate failure and recursive calling of OpenPassiveDataStream(PASV...), which then reads the successful result of the previous call to Execute("EPSV"), and fails.

**Comment by tekhedd** [comment]
Trace suggests that the culprit is my use of OpenAppend followed by "stream.Position" to detect whether I am overwriting an existing file.

                  Stream retStream = _ftpClient.OpenAppend( uri.AbsolutePath, FtpDataType.Binary );
                  try
                  {
                     if (retStream.Position > 0)
                        throw new ApplicationException("output file exists and has nonzero length");
                     // ...do more stuff and return opened stream for writing

The trace 

> AUTH TLS
> 234 Proceed with negotiation.
> Time to activate encryption: 0h 0m 0s, Total Seconds: 0.1490189.
> USER <redacted>
> 331 Please specify the password.
> PASS <omitted>
> 230 Login successful.
> PBSZ 0
> 200 PBSZ set to 0.
> PROT P
> 200 PROT now Private.
> FEAT
> 211-Features:
>  AUTH SSL
>  AUTH TLS
>  EPRT
>  EPSV
>  MDTM
>  PASV
>  PBSZ
>  PROT
>  REST STREAM
>  SIZE
>  TVFS
>  UTF8
> 211 End
> Text encoding: System.Text.UTF8Encoding
> OPTS UTF8 ON
> 200 Always in UTF8 mode.
> TYPE I
> 200 Switching to Binary mode.
> ### first transfer is successful
> SIZE /2017-05-09-16-28-21-76-296061-BOB-2016-08-25-11-53-46-66-439782.xdcf
> 550 Could not get file size.
> EPSV
> 229 Entering Extended Passive Mode (|||50028|).
> APPE /2017-05-09-16-28-21-76-296061-BOB-2016-08-25-11-53-46-66-439782.xdcf
> 150 Ok to send data.
> Time to activate encryption: 0h 0m 0s, Total Seconds: 0.0350044.
> TYPE I
> ### Maybe the problem: data left over in read buffer of socket?
> 226 Transfer complete.
> ### socket kept open for multi-file transfer. Second transfer fails.
> SIZE /2017-05-09-16-28-22-91-475437-BOB-2016-08-25-11-53-46-66-536285.xdcf
> 200 Switching to Binary mode.
> EPSV
> 550 Could not get file size.
> PASV
> 229 Entering Extended Passive Mode (|||50016|).
> ### Any other attempt will all fail similarly due to leftover data in the socket. 
> <blahblah> FluentFTP.FtpException: Malformed PASV response: Entering Extended Passive Mode (|||50016|).
> 

Is there a workaround?

**Comment by hgupta9** [comment]
Yes there is. If you are trying to append to a file, then the latest UploadFile() has an option for appending. Can you check the beta at : https://github.com/hgupta9/FluentFTP/issues/85

**Comment by hgupta9** [comment]
Also if you are using the low level methods you need to use `FtpReply status = GetReply()` after you finish transfering a file to ensure no stale data is left over.

**Comment by tekhedd** [comment]
Aha! GetReply()! I read the documentation and somehow missed that... oh it didn't exist until an hour ago! So I don't feel /quite/ so stupid now, only the usual amount. :)

Thank you for your help!

I can not use UploadFile(), as I am working with streams. (The data does not exist as files on the local filesystem, and may be transformed and compressed on the fly.) 

Note that I am actually not interested in appending, but there is no atomic way to create a new file in FTP (guaranteed without overwriting existing files). FileExists appears to be incredibly slow on the real-life servers; not a fault of the implementation, just FTP and file listsings. With the slow performance and the race condition, I concluded that I'm not gaining anything by using the FileExists check. So, I am using append (ie get size) to see whether there is an existing non-zero-length file. As long as I'm the exclusive writer in a directory it works. I have accepted the risk of overwriting 0-length files as a compromise in this application.

**Comment by tekhedd** [comment]
Um. I notice that GetReply() is protected. :o

**Comment by hgupta9** [comment]
Try these and lemme know!

1. UploadFile() supports uploading from streams. There are other overloads.
2. The performance of FileExists() has greatly improved in the latest beta. https://github.com/hgupta9/FluentFTP/issues/85
3. I've made GetReply() public in the latest beta.

**Comment by hgupta9** [comment]
I've released all these features on nuget. Can you try it? https://www.nuget.org/packages/FluentFTP/17.0.0

**Comment by tekhedd** [comment]
17.1.0's GetReply() is public and works. 

FileExists() is very efficient now thank you!

Unfortunately, the stream overloads of Upload() are not an option for me--I need a writable output Stream. The output Stream is (sometimes) wrapped with various other stream wrappers (Textwriter, XmlTextWriter etc). I would have to rewrite the output transformations as a multithreaded pipeline before I could use Upload(Stream). 

Thanks again, I am very happy!

**Comment by hgupta9** [comment]
You're welcome. And your use case for a writable output stream is very interesting. I wish something like that can be accomplished with a high level API. Or perhaps the library could auto-call GetReply() when you close the stream?

**Comment by tekhedd** [comment]
Well, most of the outputs use the (somewhat weird) WebRequest classes, which follow the form, roughly, (this isn't real code but you get the idea):

    using (Stream st = request.GetRequestStream())
    {
       GZipStream output = new GZipStream(st);
       XmlTransform.Transform(input, output);
    }

    using (WebResponse response = request.GetResponse())
    { 
        // parse response if HTTP, etc.
       if (response is HttpWebResponse) {...

So... I made a FluentFtpWebResponse, and I wrapped the FtpReply in it, so I could easily see the final code. Although it's never an error if we got this far so it's kind of pointless. :)  So yes, it's just as easy to call GetReply() in close/dispose. 

I'm thinking of how standard C streams return an error code from close(), to report errors when flushing I/O buffers. It's bad form for a stream to throw exceptions from Dispose(), IMO, but this doesn't leave many options does it, because so many IO things can go wrong. For example, this case. :) And in fact I'll bet a C-language FTP server can return an error to GetReply() for just this reason...

So, I don't know. Yes. I'm not suggesting you implement the WebRequest interface because, well, it's a bit weird and has its own flaws. Am I overanalyzing this? I'm overanalyzing this.
